src/Mvc/src/Microsoft.AspNetCore.Mvc.ViewFeatures/Infrastructure/HttpUriHelper.cs

```diff
@@ -0,0 +1,71 @@
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Extensions;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures
+{
+    internal class HttpUriHelper : UriHelperBase
+    {
+        private readonly HttpContext _context;
+
+        public HttpUriHelper(HttpContext context)
+        {
+            _context = context;
+        }
+
+        protected override void InitializeState()
+        {
+            SetAbsoluteBaseUri(GetContextBaseUri());
+            SetAbsoluteUri(GetFullUri());
+        }
+
+        private string GetFullUri()
+        {
+            var request = _context.Request;
+            return UriHelper.BuildAbsolute(
+                request.Scheme,
+                request.Host,
+                request.PathBase,
+                request.Path,
+                request.QueryString);
+        }
+
+        private string GetContextBaseUri()
+        {
+            var request = _context.Request;
+            return UriHelper.BuildAbsolute(request.Scheme, request.Host, request.PathBase);
+        }
+
+        protected override void NavigateToCore(string uri, bool forceLoad)
+        {
+            // For now throw.
+            // We don't have a good way to do redirects on a prerrendering context.
+            // Currently there's no good way to stop the rendering process. We can fix
+            // this by passing in a cancellation token to RenderRootComponentAsync and
+            // having that checked before we await on every ProcessAsynchronousWork iteration
+            // if we choose to go that way. Upon redirect we could trigger the cancellation here
+            // and do _context.Response.Redirect(uri)
+            // The bigger (not easily solvable problem) is that the response might have already
+            // started (for example when the page/view has already been flushed) which will mean
+            // that the behavior of this feature is not reliable, and not easily to diagnose as
+            // the developer only notices it when this happens at runtime; something we don't really want.
+            throw new InvalidOperationException(
+                "Redirects are not supported on a prerrendering environment.");
+
+            // We need to do something here about the hash to prevent infinite redirects
+            // as the hash portion of a url doesn't get sent to the browser.
+            // This means that the hash won't be respected for prerrendering scenarios.
+            // There are several involved things that we could do. Bake the knowledge into
+            // the IUriHelper abstraction so that we can produce a special link where we
+            // encode the hash as a query stirng parameter.
+            // We would preprocess this on the server to set the appropriate url.
+            // For example: 
+            // CurrentUrl = "http://www.localhost/base/path?query=value#top
+            // Redirect = "http://www.localhost/base/path?query=value#first
```

Comment:
I'm not sure how this scenario would occur in realistic use cases. We're not doing hash-based routing by default. Anyone who tries to use URL hashes in routing will have to deal with the fact that the server doesn't see it, just like they do in (say) MVC actions.